Importing all libraries and packages that might be used.

In [1]:
import numpy as np
from bs4 import BeautifulSoup
import requests
!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
!conda install -c conda-forge folium=0.5.0 --yes 
import folium 
import json
from pandas.io.json import json_normalize 

Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    geographiclib: 1.49-py_0   conda-forge
    geopy:         1.18.1-py_0 conda-forge

geographiclib- 100% |################################| Time: 0:00:00  23.84 MB/s
geopy-1.18.1-p 100% |################################| Time: 0:00:00  37.49 MB/s
Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    altair:  2.2.2-py35_1 conda-forge
    branca:  0.3.1-py_0   conda-forge
    folium:  0.5.0-py_0   conda-forge
    vincent: 0.4.4-py_1   conda-forge

altair-2.2.2-p 100% |################################| Time: 0:00:00  48.88 MB/s
branca-0.3.1-p 100% |################################| Time: 0:00:00  36.04 MB/s
vincent-0.4.4- 100% |###################

Adding utilities.csv file to the notebook from assets in order to work on it

In [2]:

import types
import pandas as pd
from ibm_botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

# @hidden_cell
# The following code accesses a file in your IBM Cloud Object Storage. It includes your credentials.
# You might want to remove those credentials before you share your notebook.
client_63e80d9c265f429caa11be66c2378321 = ibm_boto3.client(service_name='s3',
    ibm_api_key_id='f_xCAD3nyjl5kwhQMo0HCCg9rbdUHyxr9i5zJiNXjfGH',
    ibm_auth_endpoint="https://iam.bluemix.net/oidc/token",
    config=Config(signature_version='oauth'),
    endpoint_url='https://s3-api.us-geo.objectstorage.service.networklayer.com')

body = client_63e80d9c265f429caa11be66c2378321.get_object(Bucket='capstone-donotdelete-pr-nxgjdzwrrmtirp',Key='uscities.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df_data_1 = pd.read_csv(body)
df_data_1.head()


city      city_ascii state_id  state_name  county_fips  \
0   Prairie Ridge   Prairie Ridge       WA  Washington        53053   
1          Edison          Edison       WA  Washington        53057   
2        Packwood        Packwood       WA  Washington        53041   
3  Wautauga Beach  Wautauga Beach       WA  Washington        53035   
4          Harper          Harper       WA  Washington        53035   

  county_name      lat       lng  population  population_proper  density  \
0      Pierce  47.1443 -122.1408         NaN                NaN   1349.8   
1      Skagit  48.5602 -122.4311         NaN                NaN    127.4   
2       Lewis  46.6085 -121.6702         NaN                NaN    213.9   
3      Kitsap  47.5862 -122.5482         NaN                NaN    261.7   
4      Kitsap  47.5207 -122.5196         NaN                NaN    342.1   

    source  incorporated             timezone         zips          id  
0  polygon         False  America/Los_Angeles  98360 98391  1840037882  
1  polygon         False  America/Los_Angeles        98232  1840017314  
2  polygon         False  America/Los_Angeles        98361  1840025265  
3    point         False  America/Los_Angeles        98366  1840037725  
4    point         False  America/Los_Angeles        98366  1840037659

Filtering the dataset to keep just place and coordinate columns for Washington district in Pennysylvania

In [3]:
df=df_data_1[['city','county_name','state_name','lat','lng']].copy()
df.head()

city county_name  state_name      lat       lng
0   Prairie Ridge      Pierce  Washington  47.1443 -122.1408
1          Edison      Skagit  Washington  48.5602 -122.4311
2        Packwood       Lewis  Washington  46.6085 -121.6702
3  Wautauga Beach      Kitsap  Washington  47.5862 -122.5482
4          Harper      Kitsap  Washington  47.5207 -122.5196

In [4]:
df_PA=df[df['state_name']=='Pennsylvania'].reset_index(drop=True)
washington_data=df_PA[df_PA['county_name']=='Washington'].reset_index(drop=True)
washington_data.head()


city county_name    state_name      lat      lng
0  Cross Creek  Washington  Pennsylvania  40.3264 -80.4091
1       Thomas  Washington  Pennsylvania  40.2376 -80.0939
2     Twilight  Washington  Pennsylvania  40.1158 -79.8936
3   Canonsburg  Washington  Pennsylvania  40.2642 -80.1868
4    Atlasburg  Washington  Pennsylvania  40.3436 -80.3803

Obtaining geographical coordinates for Washington district to map it using folium.

In [5]:
address = 'Washington, PA'

geolocator = Nominatim(user_agent="PA_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Washington are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Washington are 40.1943798, -80.2541251.


In [6]:
washington_data.shape

(81, 5)

There are 81 cities in total in the Washington county.

In [7]:
# create map of Washington using latitude and longitude values
map_washington = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(washington_data['lat'], washington_data['lng'], washington_data['city']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_washington)  
    
map_washington


Passing Credentials to obtain foursquare access

In [8]:
CLIENT_ID = 'UKZ4UV54O0YKGO3WRGJ4IKPDDYAS2LRVISRLIP2IZZEX1GJS' # your Foursquare ID
CLIENT_SECRET = '2P1U5S4UBPZ2LNNIKBVQ1VZWVDTWGA5GOKT35HBQJHEJHTPS' # your Foursquare Secret
VERSION = '20190121' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ',CLIENT_ID)
print('CLIENT_SECRET:' , CLIENT_SECRET)

Your credentails:
CLIENT_ID:  UKZ4UV54O0YKGO3WRGJ4IKPDDYAS2LRVISRLIP2IZZEX1GJS
CLIENT_SECRET: 2P1U5S4UBPZ2LNNIKBVQ1VZWVDTWGA5GOKT35HBQJHEJHTPS


Getting all restaurants in the county by adding section=food

In [9]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
        LIMIT=100
        section='food'
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}&section={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT,
            section)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['city', 
                  'city Latitude', 
                  'city Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [10]:
washington_venues = getNearbyVenues(names=washington_data['city'],
                                   latitudes=washington_data['lat'],
                                   longitudes=washington_data['lng']
                                  )

Cross Creek
Thomas
Twilight
Canonsburg
Atlasburg
Primrose
Centerville
Avella
Hickory
Muse
Beallsville
Washington
Burgettstown
East Washington
McDonald
McGovern
Courtney
Finleyville
Slovan
West Middletown
Cokeburg
North Charleroi
Marianna
Elco
Ellsworth
Paris
Langeloth
Wylandville
Joffre
Wolfdale
Hendersonville
Taylorstown
Southview
Green Hills
Fisher
Bentleyville
Midway
Van Voorhis
Venice
Eighty Four
Dunlevy
Monongahela
Woodrow
Moninger
Roscoe
Murry Hill
Taylorstown Station
Wickerham Manor
Charleroi
Thompsonville
Elrama
Cecil
Manifold
California
Coal Center
Speers
Donora
Strabane
McMurray
New Eagle
Cowden
Gastonville
Meadow Lands
Candor
Millsboro
Westland
Buffalo
Fredericktown
Deemston
Claysville
Bishop
Bulger
Linden
Long Branch
Stockdale
West Alexander
Houston
West Brownsville
Lawrence
Baidland
Allenport


In [11]:
print(washington_venues.shape)
washington_venues.head()


(250, 7)


city  city Latitude  city Longitude                        Venue  \
0    Twilight        40.1158        -79.8936                   BJ's Diner   
1  Canonsburg        40.2642        -80.1868                       SUBWAY   
2  Canonsburg        40.2642        -80.1868         Little Caesars Pizza   
3  Canonsburg        40.2642        -80.1868  Big Fella's Pizza and Wings   
4  Canonsburg        40.2642        -80.1868                Vocelli Pizza   

   Venue Latitude  Venue Longitude  Venue Category  
0       40.122878       -79.890087           Diner  
1       40.260099       -80.181431  Sandwich Place  
2       40.259645       -80.182212     Pizza Place  
3       40.259135       -80.185195     Pizza Place  
4       40.258674       -80.187047     Pizza Place

So we can see there are 250 restaurants in the entire county.

In [82]:
food_group=washington_venues.groupby('city').count()
food_group=food_group.drop(['city Latitude','city Longitude','Venue Latitude','Venue Longitude','Venue Category'], axis=1)
food_group=food_group.rename(columns = {"Venue":"Food Venue"})
food_group

Food Venue
city                        
Allenport                  1
Atlasburg                  2
Avella                     2
Baidland                   2
Beallsville                1
Bentleyville               4
Bishop                     4
Burgettstown               9
Canonsburg                13
Charleroi                 10
Claysville                 4
Coal Center                6
Cowden                     5
Donora                     5
East Washington           16
Eighty Four                1
Elco                       1
Ellsworth                  4
Elrama                     3
Finleyville                4
Fisher                     1
Gastonville                6
Hendersonville             3
Hickory                    3
Houston                    7
Lawrence                   4
Manifold                  18
McDonald                   8
McMurray                   5
Meadow Lands               3
Midway                     1
Millsboro                  1
Moninger                   5
Monongahela                7
Murry Hill                10
Muse                       2
New Eagle                  3
North Charleroi            4
Paris                      1
Primrose                   1
Roscoe                     2
Speers                     4
Stockdale                  2
Strabane                   4
Thompsonville             21
Twilight                   1
Washington                20
West Alexander             2
West Brownsville           3
Wickerham Manor            1

In [13]:
print('There are {} uniques categories.'.format(len(washington_venues['Venue Category'].unique())))

There are 33 uniques categories.


There are 33 types of restaurants in the county. We create an instance for each restaurant by one hot encoding.

In [14]:
washington_onehot = pd.get_dummies(washington_venues[['Venue Category']], prefix="", prefix_sep="")

# add city column back to dataframe
washington_onehot['city'] = washington_venues['city'] 

# move city column to the first column
fixed_columns = [washington_onehot.columns[-1]] + list(washington_onehot.columns[:-1])
washington_onehot = washington_onehot[fixed_columns]

washington_onehot.head()


city  American Restaurant  Asian Restaurant  BBQ Joint  Bagel Shop  \
0    Twilight                    0                 0          0           0   
1  Canonsburg                    0                 0          0           0   
2  Canonsburg                    0                 0          0           0   
3  Canonsburg                    0                 0          0           0   
4  Canonsburg                    0                 0          0           0   

   Bakery  Breakfast Spot  Burger Joint  Café  Chinese Restaurant  \
0       0               0             0     0                   0   
1       0               0             0     0                   0   
2       0               0             0     0                   0   
3       0               0             0     0                   0   
4       0               0             0     0                   0   

      ...       New American Restaurant  Pizza Place  Restaurant  \
0     ...                             0            0           0   
1     ...                             0            0           0   
2     ...                             0            1           0   
3     ...                             0            1           0   
4     ...                             0            1           0   

   Sandwich Place  Seafood Restaurant  Snack Place  Steakhouse  \
0               0                   0            0           0   
1               1                   0            0           0   
2               0                   0            0           0   
3               0                   0            0           0   
4               0                   0            0           0   

   Sushi Restaurant  Taco Place  Wings Joint  
0                 0           0            0  
1                 0           0            0  
2                 0           0            0  
3                 0           0            0  
4                 0           0            0  

[5 rows x 34 columns]

In [15]:
washington_onehot.shape

(250, 34)

In [16]:
washington_grouped = washington_onehot.groupby('city').mean().reset_index()
washington_grouped


city  American Restaurant  Asian Restaurant  BBQ Joint  \
0          Allenport             0.000000          0.000000        0.0   
1          Atlasburg             0.000000          0.000000        0.0   
2             Avella             0.000000          0.000000        0.0   
3           Baidland             0.500000          0.000000        0.0   
4        Beallsville             0.000000          0.000000        0.0   
5       Bentleyville             0.250000          0.250000        0.0   
6             Bishop             0.000000          0.000000        0.0   
7       Burgettstown             0.111111          0.000000        0.0   
8         Canonsburg             0.000000          0.000000        0.0   
9          Charleroi             0.000000          0.000000        0.1   
10        Claysville             0.000000          0.000000        0.0   
11       Coal Center             0.333333          0.000000        0.0   
12            Cowden             0.200000          0.000000        0.0   
13            Donora             0.000000          0.000000        0.0   
14   East Washington             0.187500          0.000000        0.0   
15       Eighty Four             0.000000          0.000000        0.0   
16              Elco             0.000000          0.000000        0.0   
17         Ellsworth             0.250000          0.000000        0.0   
18            Elrama             0.000000          0.000000        0.0   
19       Finleyville             0.000000          0.000000        0.0   
20            Fisher             0.000000          0.000000        0.0   
21       Gastonville             0.166667          0.000000        0.0   
22    Hendersonville             0.333333          0.000000        0.0   
23           Hickory             0.333333          0.000000        0.0   
24           Houston             0.142857          0.000000        0.0   
25          Lawrence             0.000000          0.000000        0.0   
26          Manifold             0.111111          0.055556        0.0   
27          McDonald             0.125000          0.000000        0.0   
28          McMurray             0.000000          0.000000        0.0   
29      Meadow Lands             0.333333          0.000000        0.0   
30            Midway             0.000000          0.000000        0.0   
31         Millsboro             0.000000          0.000000        0.0   
32          Moninger             0.200000          0.000000        0.0   
33       Monongahela             0.000000          0.000000        0.0   
34        Murry Hill             0.200000          0.000000        0.0   
35              Muse             0.000000          0.000000        0.0   
36         New Eagle             0.000000          0.000000        0.0   
37   North Charleroi             0.000000          0.000000        0.0   
38             Paris             0.000000          0.000000        0.0   
39          Primrose             0.000000          0.000000        0.0   
40            Roscoe             0.000000          0.000000        0.0   
41            Speers             0.750000          0.000000        0.0   
42         Stockdale             0.000000          0.000000        0.0   
43          Strabane             0.000000          0.000000        0.0   
44     Thompsonville             0.095238          0.000000        0.0   
45          Twilight             0.000000          0.000000        0.0   
46        Washington             0.150000          0.000000        0.0   
47    West Alexander             0.500000          0.000000        0.0   
48  West Brownsville             0.000000          0.000000        0.0   
49   Wickerham Manor             0.000000          0.000000        0.0   

    Bagel Shop    Bakery  Breakfast Spot  Burger Joint      Café  \
0     0.000000  0.000000          0.0000      0.000000  0.000000   
1     0.000000  0.000000          0.0000      0.000000  0.000000   
2     0.000000  0.000000          0.0

As we see above only 50 cities out of 81 have some kind of a restaurant. 

In [17]:
num_top_venues = 34

for hood in washington_grouped['city']:
    print("----"+hood+"----")
    temp = washington_grouped[washington_grouped['city'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')


----Allenport----
                      venue  freq
0                Restaurant   1.0
1       American Restaurant   0.0
2         German Restaurant   0.0
3                Taco Place   0.0
4          Sushi Restaurant   0.0
5                Steakhouse   0.0
6               Snack Place   0.0
7        Seafood Restaurant   0.0
8            Sandwich Place   0.0
9               Pizza Place   0.0
10  New American Restaurant   0.0
11       Mexican Restaurant   0.0
12        Korean Restaurant   0.0
13       Italian Restaurant   0.0
14            Hot Dog Joint   0.0
15         Greek Restaurant   0.0
16      Fried Chicken Joint   0.0
17         Asian Restaurant   0.0
18               Food Truck   0.0
19                     Food   0.0
20     Fast Food Restaurant   0.0
21               Donut Shop   0.0
22                    Diner   0.0
23            Deli / Bodega   0.0
24  Comfort Food Restaurant   0.0
25       Chinese Restaurant   0.0
26                     Café   0.0
27             Burger Joint   

                      venue  freq
0                      Food  0.22
1       American Restaurant  0.11
2            Sandwich Place  0.11
3      Fast Food Restaurant  0.11
4               Pizza Place  0.11
5                     Diner  0.11
6        Chinese Restaurant  0.11
7   Comfort Food Restaurant  0.11
8        Seafood Restaurant  0.00
9             Hot Dog Joint  0.00
10               Restaurant  0.00
11              Snack Place  0.00
12               Steakhouse  0.00
13         Sushi Restaurant  0.00
14               Taco Place  0.00
15  New American Restaurant  0.00
16       Mexican Restaurant  0.00
17        Korean Restaurant  0.00
18       Italian Restaurant  0.00
19      Fried Chicken Joint  0.00
20         Greek Restaurant  0.00
21        German Restaurant  0.00
22         Asian Restaurant  0.00
23               Food Truck  0.00
24               Donut Shop  0.00
25            Deli / Bodega  0.00
26                     Café  0.00
27             Burger Joint  0.00
28           B

                      venue  freq
0            Sandwich Place   1.0
1       American Restaurant   0.0
2         German Restaurant   0.0
3                Taco Place   0.0
4          Sushi Restaurant   0.0
5                Steakhouse   0.0
6               Snack Place   0.0
7        Seafood Restaurant   0.0
8                Restaurant   0.0
9               Pizza Place   0.0
10  New American Restaurant   0.0
11       Mexican Restaurant   0.0
12        Korean Restaurant   0.0
13       Italian Restaurant   0.0
14            Hot Dog Joint   0.0
15         Greek Restaurant   0.0
16      Fried Chicken Joint   0.0
17         Asian Restaurant   0.0
18               Food Truck   0.0
19                     Food   0.0
20     Fast Food Restaurant   0.0
21               Donut Shop   0.0
22                    Diner   0.0
23            Deli / Bodega   0.0
24  Comfort Food Restaurant   0.0
25       Chinese Restaurant   0.0
26                     Café   0.0
27             Burger Joint   0.0
28           B

                      venue  freq
0        Italian Restaurant  0.67
1       American Restaurant  0.33
2         German Restaurant  0.00
3                Taco Place  0.00
4          Sushi Restaurant  0.00
5                Steakhouse  0.00
6               Snack Place  0.00
7        Seafood Restaurant  0.00
8            Sandwich Place  0.00
9                Restaurant  0.00
10              Pizza Place  0.00
11  New American Restaurant  0.00
12       Mexican Restaurant  0.00
13        Korean Restaurant  0.00
14            Hot Dog Joint  0.00
15         Greek Restaurant  0.00
16      Fried Chicken Joint  0.00
17         Asian Restaurant  0.00
18               Food Truck  0.00
19                     Food  0.00
20     Fast Food Restaurant  0.00
21               Donut Shop  0.00
22                    Diner  0.00
23            Deli / Bodega  0.00
24  Comfort Food Restaurant  0.00
25       Chinese Restaurant  0.00
26                     Café  0.00
27             Burger Joint  0.00
28           B

                      venue  freq
0       American Restaurant  0.33
1                Steakhouse  0.33
2                Restaurant  0.33
3         German Restaurant  0.00
4                Taco Place  0.00
5          Sushi Restaurant  0.00
6               Snack Place  0.00
7        Seafood Restaurant  0.00
8            Sandwich Place  0.00
9               Pizza Place  0.00
10  New American Restaurant  0.00
11       Mexican Restaurant  0.00
12        Korean Restaurant  0.00
13       Italian Restaurant  0.00
14            Hot Dog Joint  0.00
15         Greek Restaurant  0.00
16      Fried Chicken Joint  0.00
17         Asian Restaurant  0.00
18               Food Truck  0.00
19                     Food  0.00
20     Fast Food Restaurant  0.00
21               Donut Shop  0.00
22                    Diner  0.00
23            Deli / Bodega  0.00
24  Comfort Food Restaurant  0.00
25       Chinese Restaurant  0.00
26                     Café  0.00
27             Burger Joint  0.00
28           B

                      venue  freq
0                      Food   1.0
1       American Restaurant   0.0
2               Pizza Place   0.0
3             Hot Dog Joint   0.0
4        Italian Restaurant   0.0
5         Korean Restaurant   0.0
6        Mexican Restaurant   0.0
7   New American Restaurant   0.0
8                Restaurant   0.0
9         German Restaurant   0.0
10           Sandwich Place   0.0
11       Seafood Restaurant   0.0
12              Snack Place   0.0
13               Steakhouse   0.0
14         Sushi Restaurant   0.0
15               Taco Place   0.0
16         Greek Restaurant   0.0
17      Fried Chicken Joint   0.0
18         Asian Restaurant   0.0
19               Food Truck   0.0
20     Fast Food Restaurant   0.0
21               Donut Shop   0.0
22                    Diner   0.0
23            Deli / Bodega   0.0
24  Comfort Food Restaurant   0.0
25       Chinese Restaurant   0.0
26                     Café   0.0
27             Burger Joint   0.0
28           B

                      venue  freq
0       American Restaurant   0.5
1                Restaurant   0.5
2         German Restaurant   0.0
3                Taco Place   0.0
4          Sushi Restaurant   0.0
5                Steakhouse   0.0
6               Snack Place   0.0
7        Seafood Restaurant   0.0
8            Sandwich Place   0.0
9               Pizza Place   0.0
10  New American Restaurant   0.0
11       Mexican Restaurant   0.0
12        Korean Restaurant   0.0
13       Italian Restaurant   0.0
14            Hot Dog Joint   0.0
15         Greek Restaurant   0.0
16      Fried Chicken Joint   0.0
17         Asian Restaurant   0.0
18               Food Truck   0.0
19                     Food   0.0
20     Fast Food Restaurant   0.0
21               Donut Shop   0.0
22                    Diner   0.0
23            Deli / Bodega   0.0
24  Comfort Food Restaurant   0.0
25       Chinese Restaurant   0.0
26                     Café   0.0
27             Burger Joint   0.0
28           B

For cities that have some kind of a restaurant lets find out the least frequent types. These are the types we shall target in our strategy.

In [18]:
def return_least_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=True)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [21]:
num_top_venues = 33

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['city']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Least Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Least Common Venue'.format(ind+1))

# create a new dataframe
citys_venues_sorted = pd.DataFrame(columns=columns)
citys_venues_sorted['city'] = washington_grouped['city']

for ind in np.arange(washington_grouped.shape[0]):
    citys_venues_sorted.iloc[ind, 1:] = return_least_common_venues(washington_grouped.iloc[ind, :], num_top_venues)

citys_venues_sorted.head()


city 1st Least Common Venue 2nd Least Common Venue  \
0    Allenport    American Restaurant       Sushi Restaurant   
1    Atlasburg    American Restaurant       Sushi Restaurant   
2       Avella    American Restaurant       Sushi Restaurant   
3     Baidland    Fried Chicken Joint       Sushi Restaurant   
4  Beallsville    American Restaurant       Sushi Restaurant   

  3rd Least Common Venue 4th Least Common Venue 5th Least Common Venue  \
0             Steakhouse            Snack Place     Seafood Restaurant   
1             Steakhouse            Snack Place     Seafood Restaurant   
2             Steakhouse            Snack Place     Seafood Restaurant   
3             Steakhouse            Snack Place     Seafood Restaurant   
4             Steakhouse            Snack Place     Seafood Restaurant   

  6th Least Common Venue 7th Least Common Venue   8th Least Common Venue  \
0         Sandwich Place            Pizza Place  New American Restaurant   
1         Sandwich Place             Restaurant              Pizza Place   
2         Sandwich Place             Restaurant  New American Restaurant   
3         Sandwich Place             Restaurant              Pizza Place   
4         Sandwich Place             Restaurant              Pizza Place   

    9th Least Common Venue           ...           24th Least Common Venue  \
0       Mexican Restaurant           ...                      Burger Joint   
1  New American Restaurant           ...                        Food Truck   
2       Mexican Restaurant           ...                        Food Truck   
3  New American Restaurant           ...                              Café   
4  New American Restaurant           ...                Chinese Restaurant   

  25th Least Common Venue  26th Least Common Venue  27th Least Common Venue  \
0      Chinese Restaurant  Comfort Food Restaurant            Deli / Bodega   
1                    Café  Comfort Food Restaurant            Deli / Bodega   
2             Wings Joint       Chinese Restaurant  Comfort Food Restaurant   
3            Burger Joint           Breakfast Spot                   Bakery   
4                    Café             Burger Joint           Breakfast Spot   

  28th Least Common Venue 29th Least Common Venue 30th Least Common Venue  \
0                   Diner              Donut Shop    Fast Food Restaurant   
1                   Diner              Donut Shop    Fast Food Restaurant   
2                   Diner              Donut Shop    Fast Food Restaurant   
3              Bagel Shop               BBQ Joint        Asian Restaurant   
4              Bagel Shop               BBQ Joint        Asian Restaurant   

  31th Least Common Venue 32th Least Common Venue 33th Least Common Venue  
0                    Café             Wings Joint              Restaurant  
1      Chinese Restaurant             Wings Joint           Hot Dog Joint  
2                    Café           Deli / Bodega             Pizza Place  
3              Food Truck                   Diner     American Restaurant  
4              Food Truck             Wings Joint                  Bakery  

[5 rows x 34 columns]

We run Kmeans clustering unsupervised algorithm to divide cities into cluster areas to find out areas where maximum restaurants are clustered.

In [22]:
# set number of clusters
kclusters = 8

washington_grouped_clustering = washington_grouped.drop('city', 1)

# run k-means clustering using 350 iterations
kmeans = KMeans(n_clusters=kclusters, n_init=350, random_state=0).fit(washington_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:50]





array([0, 7, 2, 6, 5, 4, 4, 4, 4, 4, 4, 4, 4, 3, 4, 4, 1, 4, 1, 1, 2, 4, 0,
       4, 4, 0, 4, 4, 4, 0, 3, 3, 4, 4, 0, 6, 1, 4, 1, 3, 1, 0, 0, 3, 4, 6,
       4, 0, 5, 3], dtype=int32)

In [24]:
washington_merged = washington_data
washington_merged = washington_merged.join(citys_venues_sorted.set_index('city'), on='city')
# Keep only the cities that have at least 1 restaurant i.e 50 out of 81
washington_merged = washington_merged[pd.notnull(washington_merged['1st Least Common Venue'])]
# add clustering labels
washington_merged['Cluster Labels'] = kmeans.labels_

In [25]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(washington_merged['lat'], washington_merged['lng'], washington_merged['city'], washington_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [26]:
washington_merged.loc[washington_merged['Cluster Labels'] == 0, washington_merged.columns[[0] + list(range(5, washington_merged.shape[1]))]]

city 1st Least Common Venue 2nd Least Common Venue  \
2      Twilight    American Restaurant       Sushi Restaurant   
41  Monongahela    American Restaurant       Sushi Restaurant   
45   Murry Hill    Fried Chicken Joint            Snack Place   
50       Elrama    American Restaurant       Sushi Restaurant   
57     Strabane    American Restaurant       Sushi Restaurant   
69   Claysville    American Restaurant       Sushi Restaurant   
70       Bishop    American Restaurant       Sushi Restaurant   
78     Lawrence    American Restaurant       Sushi Restaurant   

   3rd Least Common Venue 4th Least Common Venue 5th Least Common Venue  \
2              Steakhouse            Snack Place     Seafood Restaurant   
41             Steakhouse            Snack Place     Seafood Restaurant   
45     Seafood Restaurant         Sandwich Place            Pizza Place   
50             Steakhouse            Snack Place     Seafood Restaurant   
57             Steakhouse            Snack Place     Seafood Restaurant   
69             Steakhouse            Snack Place     Seafood Restaurant   
70             Steakhouse     Seafood Restaurant         Sandwich Place   
78             Steakhouse            Snack Place     Seafood Restaurant   

   6th Least Common Venue   7th Least Common Venue   8th Least Common Venue  \
2          Sandwich Place               Restaurant              Pizza Place   
41             Restaurant  New American Restaurant       Mexican Restaurant   
45     Mexican Restaurant        Korean Restaurant            Hot Dog Joint   
50         Sandwich Place               Restaurant              Pizza Place   
57         Sandwich Place               Restaurant  New American Restaurant   
69         Sandwich Place               Restaurant  New American Restaurant   
70             Restaurant  New American Restaurant       Mexican Restaurant   
78         Sandwich Place               Restaurant              Pizza Place   

     9th Least Common Venue      ...        25th Least Common Venue  \
2   New American Restaurant      ...                   Burger Joint   
41        Korean Restaurant      ...                  Deli / Bodega   
45         Greek Restaurant      ...                     Bagel Shop   
50  New American Restaurant      ...                 Breakfast Spot   
57       Mexican Restaurant      ...                    Wings Joint   
69       Mexican Restaurant      ...                     Food Truck   
70            Hot Dog Joint      ...        Comfort Food Restaurant   
78  New American Restaurant      ...                    Wings Joint   

    26th Least Common Venue  27th Least Common Venue  28th Least Common Venue  \
2            Breakfast Spot                   Bakery               Bagel Shop   
41                    Diner               Donut Shop       Chinese Restaurant   
45         Sushi Restaurant               Steakhouse  New American Restaurant   
50                   Bakery               Bagel Shop                BBQ Joint   
57       Chinese Restaurant  Comfort Food Restaurant            Deli / Bodega   
69             Burger Joint           Breakfast Spot                   Bakery   
70            Deli / Bodega                    Diner                     Café   
78  Comfort Food Restaurant            Deli / Bodega                    Diner   

   29th Least Common Venue 30th Least Common Venue 31th Least Common Venue  \
2                BBQ Joint        Asian Restaurant              Food Truck   
41                    Food      Italian Restaurant          Sandwich Place   
45    Fast Food Restaurant            Burger Joint      Italian Restaurant   
50        Asian Restaurant              Food Truck             Wings Joint   
57                   Diner              Donut Shop    Fast Food Restaurant   
69              Bagel Shop               BBQ Joint                    Food   
70             Wings Joint       Korean Restaurant             Pizza Place   
78    Fast Food Restaurant      Chinese Res

In [27]:
washington_merged.loc[washington_merged['Cluster Labels'] == 1, washington_merged.columns[[0] + list(range(5, washington_merged.shape[1]))]]

city 1st Least Common Venue 2nd Least Common Venue  \
25         Paris    American Restaurant       Sushi Restaurant   
34        Fisher    American Restaurant       Sushi Restaurant   
35  Bentleyville    Fried Chicken Joint       Sushi Restaurant   
59     New Eagle    American Restaurant       Sushi Restaurant   
61   Gastonville    Fried Chicken Joint       Sushi Restaurant   
64     Millsboro    American Restaurant       Sushi Restaurant   

   3rd Least Common Venue 4th Least Common Venue 5th Least Common Venue  \
25             Steakhouse            Snack Place     Seafood Restaurant   
34             Steakhouse            Snack Place     Seafood Restaurant   
35             Steakhouse            Snack Place     Seafood Restaurant   
59             Steakhouse            Snack Place     Seafood Restaurant   
61             Steakhouse            Snack Place     Seafood Restaurant   
64             Steakhouse            Snack Place     Seafood Restaurant   

   6th Least Common Venue   7th Least Common Venue   8th Least Common Venue  \
25         Sandwich Place               Restaurant              Pizza Place   
34         Sandwich Place               Restaurant              Pizza Place   
35             Restaurant              Pizza Place  New American Restaurant   
59         Sandwich Place               Restaurant  New American Restaurant   
61             Restaurant  New American Restaurant       Mexican Restaurant   
64         Sandwich Place               Restaurant  New American Restaurant   

     9th Least Common Venue      ...       25th Least Common Venue  \
25  New American Restaurant      ...                  Burger Joint   
34  New American Restaurant      ...                  Burger Joint   
35       Mexican Restaurant      ...                    Food Truck   
59       Mexican Restaurant      ...                    Food Truck   
61        Korean Restaurant      ...                    Bagel Shop   
64       Mexican Restaurant      ...                   Wings Joint   

   26th Least Common Venue  27th Least Common Venue 28th Least Common Venue  \
25          Breakfast Spot                   Bakery              Bagel Shop   
34          Breakfast Spot                   Bakery              Bagel Shop   
35          Breakfast Spot                   Bakery              Bagel Shop   
59            Burger Joint           Breakfast Spot                  Bakery   
61          Breakfast Spot                   Bakery    Fast Food Restaurant   
64      Chinese Restaurant  Comfort Food Restaurant           Deli / Bodega   

   29th Least Common Venue 30th Least Common Venue 31th Least Common Venue  \
25               BBQ Joint        Asian Restaurant              Food Truck   
34               BBQ Joint        Asian Restaurant              Food Truck   
35               BBQ Joint        Asian Restaurant                    Food   
59              Bagel Shop               BBQ Joint        Asian Restaurant   
61             Pizza Place                    Café      Italian Restaurant   
64                   Diner              Donut Shop    Fast Food Restaurant   

   32th Least Common Venue 33th Least Common Venue Cluster Labels  
25             Wings Joint                    Food              1  
34             Wings Joint           Deli / Bodega              1  
35          Sandwich Place     American Restaurant              1  
59             Pizza Place                    Food              1  
61          Sandwich Place     American Restaurant              1  
64                    Café             Pizza Place              1  

[6 rows x 35 columns]

In [28]:
washington_merged.loc[washington_merged['Cluster Labels'] == 2, washington_merged.columns[[0] + list(range(5, washington_merged.shape[1]))]]

city 1st Least Common Venue 2nd Least Common Venue  \
4   Atlasburg    American Restaurant       Sushi Restaurant   
36     Midway    American Restaurant       Sushi Restaurant   

   3rd Least Common Venue 4th Least Common Venue 5th Least Common Venue  \
4              Steakhouse            Snack Place     Seafood Restaurant   
36             Steakhouse            Snack Place     Seafood Restaurant   

   6th Least Common Venue 7th Least Common Venue   8th Least Common Venue  \
4          Sandwich Place             Restaurant              Pizza Place   
36         Sandwich Place             Restaurant  New American Restaurant   

     9th Least Common Venue      ...       25th Least Common Venue  \
4   New American Restaurant      ...                          Café   
36       Mexican Restaurant      ...                   Wings Joint   

    26th Least Common Venue  27th Least Common Venue 28th Least Common Venue  \
4   Comfort Food Restaurant            Deli / Bodega                   Diner   
36       Chinese Restaurant  Comfort Food Restaurant           Deli / Bodega   

   29th Least Common Venue 30th Least Common Venue 31th Least Common Venue  \
4               Donut Shop    Fast Food Restaurant      Chinese Restaurant   
36                   Diner              Donut Shop    Fast Food Restaurant   

   32th Least Common Venue 33th Least Common Venue Cluster Labels  
4              Wings Joint           Hot Dog Joint              2  
36                    Café             Pizza Place              2  

[2 rows x 35 columns]

In [29]:
washington_merged.loc[washington_merged['Cluster Labels'] == 3, washington_merged.columns[[0] + list(range(5, washington_merged.shape[1]))]]

city 1st Least Common Venue 2nd Least Common Venue  \
21  North Charleroi    American Restaurant       Sushi Restaurant   
52         Manifold    Fried Chicken Joint            Pizza Place   
54      Coal Center    Fried Chicken Joint       Sushi Restaurant   
62     Meadow Lands    Fried Chicken Joint       Sushi Restaurant   
74        Stockdale    American Restaurant       Sushi Restaurant   
80        Allenport    American Restaurant       Sushi Restaurant   

   3rd Least Common Venue 4th Least Common Venue 5th Least Common Venue  \
21             Steakhouse            Snack Place     Seafood Restaurant   
52          Hot Dog Joint       Greek Restaurant      German Restaurant   
54             Steakhouse            Snack Place     Seafood Restaurant   
62            Snack Place     Seafood Restaurant         Sandwich Place   
74             Steakhouse            Snack Place     Seafood Restaurant   
80             Steakhouse            Snack Place     Seafood Restaurant   

   6th Least Common Venue   7th Least Common Venue   8th Least Common Venue  \
21         Sandwich Place  New American Restaurant       Mexican Restaurant   
52             Taco Place               Restaurant                     Food   
54             Restaurant  New American Restaurant       Mexican Restaurant   
62            Pizza Place  New American Restaurant       Mexican Restaurant   
74         Sandwich Place  New American Restaurant       Mexican Restaurant   
80         Sandwich Place              Pizza Place  New American Restaurant   

   9th Least Common Venue      ...        25th Least Common Venue  \
21      Korean Restaurant      ...        Comfort Food Restaurant   
52     Seafood Restaurant      ...             Italian Restaurant   
54      Korean Restaurant      ...                           Food   
62      Korean Restaurant      ...             Chinese Restaurant   
74      Korean Restaurant      ...                           Café   
80     Mexican Restaurant      ...             Chinese Restaurant   

    26th Least Common Venue  27th Least Common Venue 28th Least Common Venue  \
21            Deli / Bodega                    Diner              Donut Shop   
52               Food Truck         Asian Restaurant      Mexican Restaurant   
54           Breakfast Spot                BBQ Joint                  Bakery   
62  Comfort Food Restaurant            Deli / Bodega                   Diner   
74       Chinese Restaurant  Comfort Food Restaurant           Deli / Bodega   
80  Comfort Food Restaurant            Deli / Bodega                   Diner   

   29th Least Common Venue 30th Least Common Venue 31th Least Common Venue  \
21                    Café                    Food              Restaurant   
52    Fast Food Restaurant            Burger Joint     American Restaurant   
54              Bagel Shop             Pizza Place      Chinese Restaurant   
62              Donut Shop                    Café              Restaurant   
74                   Diner              Donut Shop            Burger Joint   
80              Donut Shop    Fast Food Restaurant                    Café   

   32th Least Common Venue 33th Least Common Venue Cluster Labels  
21                  Bakery             Pizza Place              3  
52          Sandwich Place              Steakhouse              3  
54          Sandwich Place     American Restaurant              3  
62              Steakhouse     American Restaurant              3  
74              Restaurant             Pizza Place              3  
80             Wings Joint              Restaurant              3  

[6 rows x 35 columns]

Cluster 4 seems to be the largest so we shall keep it in a dataframe for future use.

In [105]:
food_clustered=washington_merged.loc[washington_merged['Cluster Labels'] == 4, washington_merged.columns[[0] + list(range(5, washington_merged.shape[1]))]]
food_clustered=food_clustered.set_index('city', drop=True)
food_clustered

1st Least Common Venue 2nd Least Common Venue  \
city                                                             
Hickory             Fried Chicken Joint       Sushi Restaurant   
Muse                American Restaurant       Sushi Restaurant   
Beallsville         American Restaurant       Sushi Restaurant   
Washington          Fried Chicken Joint       Sushi Restaurant   
Burgettstown        Fried Chicken Joint       Sushi Restaurant   
East Washington             Wings Joint      Korean Restaurant   
McDonald            Fried Chicken Joint       Sushi Restaurant   
Finleyville         American Restaurant       Sushi Restaurant   
Elco                American Restaurant       Sushi Restaurant   
Ellsworth           Fried Chicken Joint       Sushi Restaurant   
Hendersonville      Fried Chicken Joint       Sushi Restaurant   
Eighty Four         American Restaurant       Sushi Restaurant   
Moninger            Fried Chicken Joint       Sushi Restaurant   
Roscoe              American Restaurant       Sushi Restaurant   
Wickerham Manor     American Restaurant       Sushi Restaurant   
Charleroi           American Restaurant       Sushi Restaurant   
Thompsonville       Fried Chicken Joint       Sushi Restaurant   
Speers              Fried Chicken Joint       Sushi Restaurant   
Donora              American Restaurant       Sushi Restaurant   
Cowden              Fried Chicken Joint       Sushi Restaurant   
West Alexander      Fried Chicken Joint       Sushi Restaurant   
West Brownsville    American Restaurant       Sushi Restaurant   

                 3rd Least Common Venue 4th Least Common Venue  \
city                                                             
Hickory                      Steakhouse            Snack Place   
Muse                         Steakhouse            Snack Place   
Beallsville                  Steakhouse            Snack Place   
Washington                   Steakhouse            Snack Place   
Burgettstown                 Steakhouse            Snack Place   
East Washington      Italian Restaurant          Hot Dog Joint   
McDonald                     Steakhouse            Snack Place   
Finleyville                  Steakhouse            Snack Place   
Elco                         Steakhouse            Snack Place   
Ellsworth                    Steakhouse            Snack Place   
Hendersonville               Steakhouse            Snack Place   
Eighty Four                  Steakhouse            Snack Place   
Moninger                     Steakhouse            Snack Place   
Roscoe                       Steakhouse            Snack Place   
Wickerham Manor              Steakhouse            Snack Place   
Charleroi                    Steakhouse            Snack Place   
Thompsonville                Steakhouse            Snack Place   
Speers                       Steakhouse            Snack Place   
Donora                       Steakhouse            Snack Place   
Cowden                       Steakhouse            Snack Place   
West Alexander               Steakhouse            Snack Place   
West Brownsville             Steakhouse            Snack Place   

                   5th Least Common Venue   6th Least Common Venue  \
city                                                                 
Hickory                Seafood Restaurant           Sandwich Place   
Muse                   Seafood Restaurant           Sandwich Place   
Beallsville            Seafood Restaurant           Sandwich Place   
Washington        New American Restaurant       Mexican Restaurant   
Burgettstown           Seafood Restaurant               Restaurant   
East Washington          Greek Restaurant        German Restaurant   
McDonald               Seafood Restaurant  New American Restaurant   
Finleyville            Seafood Restaurant               Restaurant   
Elco                   Seafood Restaurant           Sandwich Place   
Ellsworth              Seafood Restaurant           Sandwich Place   
Henders

In [31]:
washington_merged.loc[washington_merged['Cluster Labels'] == 5, washington_merged.columns[[0] + list(range(5, washington_merged.shape[1]))]]                            
                             

city 1st Least Common Venue 2nd Least Common Venue  \
7     Avella    American Restaurant       Sushi Restaurant   
79  Baidland    Fried Chicken Joint       Sushi Restaurant   

   3rd Least Common Venue 4th Least Common Venue 5th Least Common Venue  \
7              Steakhouse            Snack Place     Seafood Restaurant   
79             Steakhouse            Snack Place     Seafood Restaurant   

   6th Least Common Venue 7th Least Common Venue   8th Least Common Venue  \
7          Sandwich Place             Restaurant  New American Restaurant   
79         Sandwich Place             Restaurant              Pizza Place   

     9th Least Common Venue      ...       25th Least Common Venue  \
7        Mexican Restaurant      ...                   Wings Joint   
79  New American Restaurant      ...                  Burger Joint   

   26th Least Common Venue  27th Least Common Venue 28th Least Common Venue  \
7       Chinese Restaurant  Comfort Food Restaurant                   Diner   
79          Breakfast Spot                   Bakery              Bagel Shop   

   29th Least Common Venue 30th Least Common Venue 31th Least Common Venue  \
7               Donut Shop    Fast Food Restaurant                    Café   
79               BBQ Joint        Asian Restaurant              Food Truck   

   32th Least Common Venue 33th Least Common Venue Cluster Labels  
7            Deli / Bodega             Pizza Place              5  
79                   Diner     American Restaurant              5  

[2 rows x 35 columns]

In [32]:
washington_merged.loc[washington_merged['Cluster Labels'] == 5, washington_merged.columns[[0] + list(range(5, washington_merged.shape[1]))]]

city 1st Least Common Venue 2nd Least Common Venue  \
7     Avella    American Restaurant       Sushi Restaurant   
79  Baidland    Fried Chicken Joint       Sushi Restaurant   

   3rd Least Common Venue 4th Least Common Venue 5th Least Common Venue  \
7              Steakhouse            Snack Place     Seafood Restaurant   
79             Steakhouse            Snack Place     Seafood Restaurant   

   6th Least Common Venue 7th Least Common Venue   8th Least Common Venue  \
7          Sandwich Place             Restaurant  New American Restaurant   
79         Sandwich Place             Restaurant              Pizza Place   

     9th Least Common Venue      ...       25th Least Common Venue  \
7        Mexican Restaurant      ...                   Wings Joint   
79  New American Restaurant      ...                  Burger Joint   

   26th Least Common Venue  27th Least Common Venue 28th Least Common Venue  \
7       Chinese Restaurant  Comfort Food Restaurant                   Diner   
79          Breakfast Spot                   Bakery              Bagel Shop   

   29th Least Common Venue 30th Least Common Venue 31th Least Common Venue  \
7               Donut Shop    Fast Food Restaurant                    Café   
79               BBQ Joint        Asian Restaurant              Food Truck   

   32th Least Common Venue 33th Least Common Venue Cluster Labels  
7            Deli / Bodega             Pizza Place              5  
79                   Diner     American Restaurant              5  

[2 rows x 35 columns]

In [33]:
washington_merged.loc[washington_merged['Cluster Labels'] == 7, washington_merged.columns[[0] + list(range(5, washington_merged.shape[1]))]]

city 1st Least Common Venue 2nd Least Common Venue  \
3  Canonsburg    American Restaurant       Sushi Restaurant   

  3rd Least Common Venue 4th Least Common Venue 5th Least Common Venue  \
3             Steakhouse            Snack Place     Seafood Restaurant   

  6th Least Common Venue   7th Least Common Venue 8th Least Common Venue  \
3             Restaurant  New American Restaurant     Mexican Restaurant   

  9th Least Common Venue      ...       25th Least Common Venue  \
3      Korean Restaurant      ...                 Deli / Bodega   

  26th Least Common Venue 27th Least Common Venue 28th Least Common Venue  \
3              Donut Shop      Chinese Restaurant             Wings Joint   

  29th Least Common Venue 30th Least Common Venue 31th Least Common Venue  \
3        Greek Restaurant                   Diner          Sandwich Place   

  32th Least Common Venue 33th Least Common Venue Cluster Labels  
3                    Food             Pizza Place              7  

[1 rows x 35 columns]

In [34]:
washington_merged.loc[washington_merged['Cluster Labels'] == 8, washington_merged.columns[[0] + list(range(5, washington_merged.shape[1]))]]

Empty DataFrame
Columns: [city, 1st Least Common Venue, 2nd Least Common Venue, 3rd Least Common Venue, 4th Least Common Venue, 5th Least Common Venue, 6th Least Common Venue, 7th Least Common Venue, 8th Least Common Venue, 9th Least Common Venue, 10th Least Common Venue, 11th Least Common Venue, 12th Least Common Venue, 13th Least Common Venue, 14th Least Common Venue, 15th Least Common Venue, 16th Least Common Venue, 17th Least Common Venue, 18th Least Common Venue, 19th Least Common Venue, 20th Least Common Venue, 21th Least Common Venue, 22th Least Common Venue, 23th Least Common Venue, 24th Least Common Venue, 25th Least Common Venue, 26th Least Common Venue, 27th Least Common Venue, 28th Least Common Venue, 29th Least Common Venue, 30th Least Common Venue, 31th Least Common Venue, 32th Least Common Venue, 33th Least Common Venue, Cluster Labels]
Index: []

[0 rows x 35 columns]

Then we shall get the shopping areas in Washington county by using section= shops.

In [35]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    shops_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
        LIMIT=100
        section='shops'
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}&section={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT,
            section)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        shops_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_shops = pd.DataFrame([item for shop_list in shops_list for item in shop_list])
    nearby_shops.columns = ['city', 
                  'city Latitude', 
                  'city Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_shops)


In [36]:
washington_shops = getNearbyVenues(names=washington_data['city'],
                                   latitudes=washington_data['lat'],
                                   longitudes=washington_data['lng']
                                  )

Cross Creek
Thomas
Twilight
Canonsburg
Atlasburg
Primrose
Centerville
Avella
Hickory
Muse
Beallsville
Washington
Burgettstown
East Washington
McDonald
McGovern
Courtney
Finleyville
Slovan
West Middletown
Cokeburg
North Charleroi
Marianna
Elco
Ellsworth
Paris
Langeloth
Wylandville
Joffre
Wolfdale
Hendersonville
Taylorstown
Southview
Green Hills
Fisher
Bentleyville
Midway
Van Voorhis
Venice
Eighty Four
Dunlevy
Monongahela
Woodrow
Moninger
Roscoe
Murry Hill
Taylorstown Station
Wickerham Manor
Charleroi
Thompsonville
Elrama
Cecil
Manifold
California
Coal Center
Speers
Donora
Strabane
McMurray
New Eagle
Cowden
Gastonville
Meadow Lands
Candor
Millsboro
Westland
Buffalo
Fredericktown
Deemston
Claysville
Bishop
Bulger
Linden
Long Branch
Stockdale
West Alexander
Houston
West Brownsville
Lawrence
Baidland
Allenport


In [37]:
print(washington_shops.shape)
washington_shops.head()


(487, 7)


city  city Latitude  city Longitude                         Venue  \
0      Thomas        40.2376        -80.0939           Pine Ridge Minimart   
1      Thomas        40.2376        -80.0939         Re-Bath of Pittsburgh   
2      Thomas        40.2376        -80.0939                ABH Supply LLC   
3      Thomas        40.2376        -80.0939  ABH Stair & Millwork Company   
4  Canonsburg        40.2642        -80.1868                 Ross Beverage   

   Venue Latitude  Venue Longitude                 Venue Category  
0       40.238685       -80.094881              Convenience Store  
1       40.240303       -80.095882                 Shop & Service  
2       40.237849       -80.089806               Business Service  
3       40.236371       -80.087908  Paper / Office Supplies Store  
4       40.259916       -80.181603                   Liquor Store

There are 487 shopping centers in Washington county

In [81]:
shop_group=washington_shops.groupby('city').count()
shop_group=shop_group.drop(['city Latitude','city Longitude','Venue Latitude','Venue Longitude','Venue Category'], axis=1)
shop_group=shop_group.rename(columns = {"Venue":"Shop Venue"})
shop_group #This creates a dataframe showing number of shopping venues in each city

Shop Venue
city                           
Atlasburg                     3
Avella                        2
Beallsville                   1
Bentleyville                  6
Bishop                        4
Buffalo                       1
Bulger                        2
Burgettstown                 10
Canonsburg                   32
Cecil                         2
Charleroi                    23
Claysville                    5
Coal Center                   5
Cokeburg                      2
Cowden                        5
Donora                        7
Dunlevy                       2
East Washington              24
Eighty Four                   4
Elco                          1
Ellsworth                     1
Elrama                        3
Finleyville                  11
Fisher                        5
Fredericktown                 3
Gastonville                   4
Hendersonville                5
Hickory                       4
Houston                      27
Joffre                        1
Lawrence                      4
Manifold                     19
McDonald                     13
McGovern                      3
McMurray                     11
Meadow Lands                  4
Midway                        3
Millsboro                     3
Moninger                     11
Monongahela                  22
Murry Hill                   16
New Eagle                     7
North Charleroi               4
Roscoe                        4
Slovan                        1
Speers                        2
Stockdale                     1
Strabane                     19
Taylorstown                   1
Taylorstown Station           1
Thomas                        4
Thompsonville                34
Venice                        1
Washington                   77
West Alexander                2
West Brownsville              5
Wickerham Manor               6
Wolfdale                      4

In [39]:
shops_onehot = pd.get_dummies(washington_shops[['Venue Category']], prefix="", prefix_sep="")

# add city column back to dataframe
shops_onehot['city'] = washington_shops['city'] 

# move city column to the first column
fixed_columns = [shops_onehot.columns[-1]] + list(shops_onehot.columns[:-1])
shops_onehot = shops_onehot[fixed_columns]

shops_onehot.head()

city  ATM  Antique Shop  Arts & Crafts Store  Auto Garage  \
0      Thomas    0             0                    0            0   
1      Thomas    0             0                    0            0   
2      Thomas    0             0                    0            0   
3      Thomas    0             0                    0            0   
4  Canonsburg    0             0                    0            0   

   Auto Workshop  Beer Store  Big Box Store  Bookstore  Boutique  \
0              0           0              0          0         0   
1              0           0              0          0         0   
2              0           0              0          0         0   
3              0           0              0          0         0   
4              0           0              0          0         0   

        ...         Shoe Store  Shop & Service  Shopping Mall  Smoke Shop  \
0       ...                  0               0              0           0   
1       ...                  0               1              0           0   
2       ...                  0               0              0           0   
3       ...                  0               0              0           0   
4       ...                  0               0              0           0   

   Smoothie Shop  Sporting Goods Shop  Supermarket  Thrift / Vintage Store  \
0              0                    0            0                       0   
1              0                    0            0                       0   
2              0                    0            0                       0   
3              0                    0            0                       0   
4              0                    0            0                       0   

   Video Store  Warehouse Store  
0            0                0  
1            0                0  
2            0                0  
3            0                0  
4            0                0  

[5 rows x 76 columns]

Inspection of dataframe shows that there are 76 types of shopping centers. Most cities contain about 10 or less types of shops. So instead of dealing with bulky dataframe with lots if negative values for most shops we shall just filter it to keep the 10 most frequent types of shopping places in each city.

In [40]:
shops_grouped = shops_onehot.groupby('city').mean().reset_index()
shops_grouped.head()

city  ATM  Antique Shop  Arts & Crafts Store  Auto Garage  \
0     Atlasburg  0.0           0.0                  0.0          0.0   
1        Avella  0.0           0.0                  0.0          0.0   
2   Beallsville  0.0           0.0                  0.0          0.0   
3  Bentleyville  0.0           0.0                  0.0          0.0   
4        Bishop  0.0           0.0                  0.0          0.0   

   Auto Workshop  Beer Store  Big Box Store  Bookstore  Boutique  \
0            0.0         0.0            0.0        0.0       0.0   
1            0.0         0.0            0.0        0.0       0.0   
2            0.0         0.0            0.0        0.0       0.0   
3            0.0         0.0            0.0        0.0       0.0   
4            0.0         0.0            0.0        0.0       0.0   

        ...         Shoe Store  Shop & Service  Shopping Mall  Smoke Shop  \
0       ...           0.333333             0.0            0.0         0.0   
1       ...           0.000000             0.0            0.0         0.0   
2       ...           0.000000             0.0            0.0         0.0   
3       ...           0.000000             0.0            0.0         0.0   
4       ...           0.000000             0.0            0.0         0.0   

   Smoothie Shop  Sporting Goods Shop  Supermarket  Thrift / Vintage Store  \
0            0.0                  0.0          0.0                     0.0   
1            0.0                  0.0          0.0                     0.0   
2            0.0                  0.0          0.0                     0.0   
3            0.0                  0.0          0.0                     0.0   
4            0.0                  0.0          0.0                     0.0   

   Video Store  Warehouse Store  
0     0.000000              0.0  
1     0.000000              0.0  
2     0.000000              0.0  
3     0.166667              0.0  
4     0.000000              0.0  

[5 rows x 76 columns]

In [41]:
num_top_venues = 10

for hood in shops_grouped['city']:
    print("----"+hood+"----")
    temp = shops_grouped[shops_grouped['city'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Atlasburg----
                  venue  freq
0     Food & Drink Shop  0.33
1            Shoe Store  0.33
2        Farmers Market  0.33
3           Music Store  0.00
4       Motorcycle Shop  0.00
5     Mobile Phone Shop  0.00
6        Massage Studio  0.00
7                Market  0.00
8  Marijuana Dispensary  0.00
9                   ATM  0.00


----Avella----
                           venue  freq
0               Business Service   0.5
1     Construction & Landscaping   0.5
2  Paper / Office Supplies Store   0.0
3              Other Repair Shop   0.0
4                   Optical Shop   0.0
5                    Music Store   0.0
6                Motorcycle Shop   0.0
7              Mobile Phone Shop   0.0
8                 Massage Studio   0.0
9                         Market   0.0


----Beallsville----
                  venue  freq
0          Home Service   1.0
1  Marijuana Dispensary   0.0
2     Other Repair Shop   0.0
3          Optical Shop   0.0
4           Music Store   0.0
5   

                    venue  freq
0        Business Service   0.2
1             Video Store   0.2
2             Supermarket   0.2
3                Pharmacy   0.2
4  Furniture / Home Store   0.2
5               Locksmith   0.0
6             Music Store   0.0
7         Motorcycle Shop   0.0
8       Mobile Phone Shop   0.0
9          Massage Studio   0.0


----Fredericktown----
                  venue  freq
0            Shoe Store  0.33
1     Convenience Store  0.33
2          Liquor Store  0.33
3  Marijuana Dispensary  0.00
4          Optical Shop  0.00
5           Music Store  0.00
6       Motorcycle Shop  0.00
7     Mobile Phone Shop  0.00
8        Massage Studio  0.00
9                Market  0.00


----Gastonville----
               venue  freq
0   Business Service  0.25
1     Discount Store  0.25
2  Other Repair Shop  0.25
3        Flower Shop  0.25
4       Optical Shop  0.00
5        Music Store  0.00
6    Motorcycle Shop  0.00
7  Mobile Phone Shop  0.00
8     Massage Studio  0.00
9 

                        venue  freq
0  Construction & Landscaping   0.5
1                   Gift Shop   0.5
2                         ATM   0.0
3        Marijuana Dispensary   0.0
4                Optical Shop   0.0
5                 Music Store   0.0
6             Motorcycle Shop   0.0
7           Mobile Phone Shop   0.0
8              Massage Studio   0.0
9                      Market   0.0


----Stockdale----
                  venue  freq
0        Clothing Store   1.0
1                   ATM   0.0
2  Marijuana Dispensary   0.0
3          Optical Shop   0.0
4           Music Store   0.0
5       Motorcycle Shop   0.0
6     Mobile Phone Shop   0.0
7        Massage Studio   0.0
8                Market   0.0
9             Locksmith   0.0


----Strabane----
                        venue  freq
0  Construction & Landscaping  0.26
1               Auto Workshop  0.11
2                Home Service  0.11
3              Discount Store  0.05
4                Carpet Store  0.05
5             Motor

In [42]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [43]:
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['city']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
citys_shops_sorted = pd.DataFrame(columns=columns)
citys_shops_sorted['city'] = shops_grouped['city']

for ind in np.arange(shops_grouped.shape[0]):
    citys_shops_sorted.iloc[ind, 1:] = return_most_common_venues(shops_grouped.iloc[ind, :], num_top_venues)

citys_shops_sorted.head()

city       1st Most Common Venue 2nd Most Common Venue  \
0     Atlasburg                  Shoe Store     Food & Drink Shop   
1        Avella  Construction & Landscaping      Business Service   
2   Beallsville                Home Service       Warehouse Store   
3  Bentleyville           Other Repair Shop           Video Store   
4        Bishop  Construction & Landscaping      Business Service   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0        Farmers Market           Flea Market        Discount Store   
1       Warehouse Store        Farmers Market             Drugstore   
2           Flea Market        Discount Store             Drugstore   
3          Home Service        Discount Store           IT Services   
4       Warehouse Store        Farmers Market             Drugstore   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0             Drugstore           Dry Cleaner     Electronics Store   
1           Dry Cleaner     Electronics Store         Event Service   
2           Dry Cleaner     Electronics Store         Event Service   
3              Pharmacy       Warehouse Store           Fabric Shop   
4           Dry Cleaner     Electronics Store         Event Service   

  9th Most Common Venue 10th Most Common Venue  
0         Event Service            Fabric Shop  
1           Fabric Shop            Flea Market  
2           Fabric Shop         Farmers Market  
3             Drugstore            Dry Cleaner  
4           Fabric Shop            Flea Market

We shall divide the county into clusters of areas with similar shopping centers.

In [44]:
# set number of clusters
kclusters = 8

shops_grouped_clustering = shops_grouped.drop('city', 1)

# run k-means clustering with 350 iterations
kmeans = KMeans(n_clusters=kclusters, n_init=350, random_state=0).fit(shops_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 3, 7, 0, 3, 3, 2, 0, 0, 7], dtype=int32)

In [45]:
shops_merged = washington_data
shops_merged = shops_merged.join(citys_shops_sorted.set_index('city'), on='city')
shops_merged = shops_merged[pd.notnull(shops_merged['1st Most Common Venue'])]
# add clustering labels
shops_merged['Cluster Labels'] = kmeans.labels_

In [46]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(shops_merged['lat'], shops_merged['lng'], shops_merged['city'], shops_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [106]:
shops_clustered=shops_merged.loc[shops_merged['Cluster Labels'] == 0, shops_merged.columns[[0] + list(range(5, shops_merged.shape[1]))]]
shops_clustered=shops_clustered.set_index('city',drop=True)
shops_clustered #Since cluster 1 seems to have maximum cities with shopping areas we shall make a dataframe from it

1st Most Common Venue  \
city                                              
Thomas                        Convenience Store   
Avella               Construction & Landscaping   
Burgettstown                   Business Service   
East Washington                          Lawyer   
McGovern                Health & Beauty Service   
Finleyville          Construction & Landscaping   
Slovan                        Other Repair Shop   
Cokeburg                      Electronics Store   
North Charleroi               Other Repair Shop   
Elco                           Business Service   
Ellsworth            Construction & Landscaping   
Joffre                         Business Service   
Wolfdale             Construction & Landscaping   
Bentleyville                  Other Repair Shop   
Midway                         Business Service   
Venice                           Discount Store   
Eighty Four                    Business Service   
Dunlevy                       Electronics Store   
Monongahela                    Business Service   
Moninger                       Business Service   
Taylorstown Station              Rental Service   
Wickerham Manor                Business Service   
Charleroi                      Business Service   
Thompsonville                         Pet Store   
Elrama                      Arts & Crafts Store   
Coal Center                    Business Service   
Speers               Construction & Landscaping   
Donora                         Business Service   
Strabane             Construction & Landscaping   
McMurray             Construction & Landscaping   
New Eagle                     Other Repair Shop   
Millsboro                         Grocery Store   
Claysville                       Discount Store   
Bishop               Construction & Landscaping   
Stockdale                        Clothing Store   
Houston                        Business Service   
West Brownsville                   Liquor Store   

                             2nd Most Common Venue  \
city                                                 
Thomas               Paper / Office Supplies Store   
Avella                            Business Service   
Burgettstown                              Pharmacy   
East Washington                        IT Services   
McGovern                Construction & Landscaping   
Finleyville                                    ATM   
Slovan                              Farmers Market   
Cokeburg                             Event Service   
North Charleroi                  Convenience Store   
Elco                               Warehouse Store   
Ellsworth                          Warehouse Store   
Joffre                             Warehouse Store   
Wolfdale                         Other Repair Shop   
Bentleyville                           Video Store   
Midway                               Auto Workshop   
Venice                             Warehouse Store   
Eighty Four                           Home Service   
Dunlevy                          Food & Drink Shop   
Monongahela                 Thrift / Vintage Store   
Moninger                   Health & Beauty Service   
Taylorstown Station                Warehouse Store   
Wickerham Manor                        Video Store   
Charleroi               Construction & Landscaping   
Thompsonville           Construction & Landscaping   
Elrama                  Construction & Landscaping   
Coal Center                            Flower Shop   
Speers                                   Gift Shop   
Donora                              Discount Store   
Strabane                             Auto Workshop   
McMurray                          Business Service   
New Eagle                      Arts & Crafts Store   
Millsboro                           Discount Store   
Claysville                        Business Service   
Bishop                            Business Service   
Stockdale                          Warehouse Store   
Houston                          Elect

In [48]:
shops_merged.loc[shops_merged['Cluster Labels'] == 1, shops_merged.columns[[0] + list(range(5, shops_merged.shape[1]))]]

city 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
60  Cowden      Business Service          Home Service     Convenience Store   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
60      Insurance Office       Warehouse Store             Drugstore   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
60           Dry Cleaner     Electronics Store         Event Service   

   10th Most Common Venue  Cluster Labels  
60            Fabric Shop               1

In [49]:
shops_merged.loc[shops_merged['Cluster Labels'] == 2, shops_merged.columns[[0] + list(range(5, shops_merged.shape[1]))]]

city 1st Most Common Venue       2nd Most Common Venue  \
11      Washington                Lawyer                Home Service   
30  Hendersonville   Sporting Goods Shop      Furniture / Home Store   
44          Roscoe        Discount Store           Convenience Store   
51           Cecil          Home Service  Construction & Landscaping   

   3rd Most Common Venue       4th Most Common Venue 5th Most Common Venue  \
11      Business Service  Construction & Landscaping              Pharmacy   
30             Pet Store              Shop & Service              Boutique   
44        Clothing Store            Business Service      Department Store   
51       Warehouse Store                 Flea Market             Drugstore   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11          Liquor Store     Convenience Store           IT Services   
30       Warehouse Store           Fabric Shop        Discount Store   
44             Drugstore           Dry Cleaner     Electronics Store   
51           Dry Cleaner     Electronics Store         Event Service   

   9th Most Common Venue 10th Most Common Venue  Cluster Labels  
11      Department Store         Discount Store               2  
30             Drugstore            Dry Cleaner               2  
44         Event Service            Fabric Shop               2  
51           Fabric Shop         Farmers Market               2

In [50]:
shops_merged.loc[shops_merged['Cluster Labels'] == 3, shops_merged.columns[[0] + list(range(5, shops_merged.shape[1]))]]

city       1st Most Common Venue       2nd Most Common Venue  \
3       Canonsburg  Construction & Landscaping            Business Service   
8          Hickory            Business Service                Home Service   
10     Beallsville                Home Service             Warehouse Store   
31     Taylorstown              Rental Service             Warehouse Store   
34          Fisher            Business Service                 Video Store   
45      Murry Hill                      Lawyer  Construction & Landscaping   
61     Gastonville           Other Repair Shop              Discount Store   
75  West Alexander              Lighting Store  Construction & Landscaping   
78        Lawrence  Construction & Landscaping         Arts & Crafts Store   

     3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
3   Furniture / Home Store             Flower Shop           Auto Garage   
8         Insurance Office      Child Care Service       Warehouse Store   
10             Flea Market          Discount Store             Drugstore   
31          Farmers Market          Discount Store             Drugstore   
34             Supermarket  Furniture / Home Store              Pharmacy   
45           Event Service        Business Service                   ATM   
61             Flower Shop        Business Service        Farmers Market   
75             Flea Market               Drugstore           Dry Cleaner   
78         Warehouse Store             Flea Market             Drugstore   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
3           Liquor Store     Mobile Phone Shop          Home Service   
8              Drugstore           Dry Cleaner     Electronics Store   
10           Dry Cleaner     Electronics Store         Event Service   
31           Dry Cleaner     Electronics Store         Event Service   
34       Warehouse Store        Discount Store             Drugstore   
45      Insurance Office           Dry Cleaner        Rental Service   
61             Drugstore           Dry Cleaner     Electronics Store   
75     Electronics Store         Event Service           Fabric Shop   
78           Dry Cleaner     Electronics Store         Event Service   

   9th Most Common Venue 10th Most Common Venue  Cluster Labels  
3            Dry Cleaner                 Lawyer               3  
8          Event Service            Fabric Shop               3  
10           Fabric Shop         Farmers Market               3  
31           Fabric Shop            Flea Market               3  
34           Dry Cleaner      Electronics Store               3  
45         Shopping Mall           Home Service               3  
61         Event Service            Fabric Shop               3  
75        Farmers Market        Warehouse Store               3  
78           Fabric Shop         Farmers Market               3

In [51]:
shops_merged.loc[shops_merged['Cluster Labels'] == 4, shops_merged.columns[[0] + list(range(5, shops_merged.shape[1]))]]

city 1st Most Common Venue 2nd Most Common Venue  \
62  Meadow Lands      Business Service          Home Service   

   3rd Most Common Venue       4th Most Common Venue 5th Most Common Venue  \
62             Gift Shop  Construction & Landscaping       Warehouse Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
62             Drugstore           Dry Cleaner     Electronics Store   

   9th Most Common Venue 10th Most Common Venue  Cluster Labels  
62         Event Service            Fabric Shop               4

In [52]:
shops_merged.loc[shops_merged['Cluster Labels'] == 5, shops_merged.columns[[0] + list(range(5, shops_merged.shape[1]))]]

city       1st Most Common Venue 2nd Most Common Venue  \
66        Buffalo  Construction & Landscaping       Warehouse Store   
67  Fredericktown                Liquor Store     Convenience Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
66        Farmers Market        Discount Store             Drugstore   
67            Shoe Store       Warehouse Store        Discount Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
66           Dry Cleaner     Electronics Store         Event Service   
67             Drugstore           Dry Cleaner     Electronics Store   

   9th Most Common Venue 10th Most Common Venue  Cluster Labels  
66           Fabric Shop            Flea Market               5  
67         Event Service            Fabric Shop               5

In [56]:
shops_merged.loc[shops_merged['Cluster Labels'] == 6, shops_merged.columns[[0] + list(range(5, shops_merged.shape[1]))]]


city 1st Most Common Venue 2nd Most Common Venue  \
52  Manifold   Sporting Goods Shop         Grocery Store   
71    Bulger      Business Service       Warehouse Store   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
52  Furniture / Home Store        Discount Store            Hobby Shop   
71          Farmers Market        Discount Store             Drugstore   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
52            Print Shop           Fabric Shop           Dry Cleaner   
71           Dry Cleaner     Electronics Store         Event Service   

   9th Most Common Venue   10th Most Common Venue  Cluster Labels  
52      Department Store  Health & Beauty Service               6  
71           Fabric Shop              Flea Market               6

In [54]:
shops_merged.loc[shops_merged['Cluster Labels'] == 7, shops_merged.columns[[0] + list(range(5, shops_merged.shape[1]))]]

city 1st Most Common Venue 2nd Most Common Venue  \
4   Atlasburg            Shoe Store     Food & Drink Shop   
14   McDonald              Pharmacy        Discount Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4         Farmers Market           Flea Market        Discount Store   
14      Business Service           Supermarket   Sporting Goods Shop   

   6th Most Common Venue 7th Most Common Venue       8th Most Common Venue  \
4              Drugstore           Dry Cleaner           Electronics Store   
14     Health Food Store           Video Store  Construction & Landscaping   

   9th Most Common Venue 10th Most Common Venue  Cluster Labels  
4          Event Service            Fabric Shop               7  
14      Insurance Office      Mobile Phone Shop               7

We shall find out the cities that have both high number of shopping areas as well as high number of restaurants. We shall also include in the dataframe the absolute number of shops and restarants in the city. This we shall acheive by intersection of the most frequent cluster dataframes and then adding on the absolute number of shops/restaurants dataframes. We shall sum the absolute number of shops and restaurants to find out the absolute number of "Hip Places" in the cities of intersection

In [127]:
washington_hip= pd.concat ([shops_clustered, food_clustered,shop_group,food_group],axis=1)
washington_hip= washington_hip[pd.notnull(washington_hip['1st Least Common Venue'])]
washington_hip= washington_hip[pd.notnull(washington_hip['1st Most Common Venue'])]
washington_hip= washington_hip[pd.notnull(washington_hip['Shop Venue'])]
washington_hip= washington_hip[pd.notnull(washington_hip['Food Venue'])]
washington_hip['Hip Places'] = washington_hip[['Shop Venue', 'Food Venue']].sum(axis=1)
washington_hip.head()

1st Most Common Venue       2nd Most Common Venue  \
Burgettstown         Business Service                    Pharmacy   
Charleroi            Business Service  Construction & Landscaping   
Donora               Business Service              Discount Store   
East Washington                Lawyer                 IT Services   
Eighty Four          Business Service                Home Service   

                3rd Most Common Venue 4th Most Common Venue  \
Burgettstown            Grocery Store          Optical Shop   
Charleroi                Home Service      Recording Studio   
Donora                  Grocery Store     Convenience Store   
East Washington      Business Service        Rental Service   
Eighty Four                    Lawyer       Warehouse Store   

                  5th Most Common Venue 6th Most Common Venue  \
Burgettstown             Discount Store        Clothing Store   
Charleroi        Furniture / Home Store              Pharmacy   
Donora                    Auto Workshop              Pharmacy   
East Washington            Costume Shop        Discount Store   
Eighty Four                   Drugstore           Dry Cleaner   

                7th Most Common Venue 8th Most Common Venue  \
Burgettstown                Pet Store         Event Service   
Charleroi              Discount Store     Mobile Phone Shop   
Donora                 Farmers Market             Drugstore   
East Washington     Convenience Store            Laundromat   
Eighty Four         Electronics Store         Event Service   

                  9th Most Common Venue 10th Most Common Venue    ...      \
Burgettstown                  Drugstore            Dry Cleaner    ...       
Charleroi        Thrift / Vintage Store                 Lawyer    ...       
Donora                      Dry Cleaner      Electronics Store    ...       
East Washington      Photography Studio             Print Shop    ...       
Eighty Four                 Fabric Shop         Farmers Market    ...       

                 28th Least Common Venue  29th Least Common Venue  \
Burgettstown          Chinese Restaurant  Comfort Food Restaurant   
Charleroi                    Pizza Place           Breakfast Spot   
Donora                            Bakery               Bagel Shop   
East Washington     Fast Food Restaurant                    Diner   
Eighty Four                        Diner               Donut Shop   

                30th Least Common Venue 31th Least Common Venue  \
Burgettstown                      Diner          Sandwich Place   
Charleroi                Sandwich Place               BBQ Joint   
Donora                        BBQ Joint        Asian Restaurant   
East Washington          Breakfast Spot              Donut Shop   
Eighty Four        Fast Food Restaurant                    Café   

                32th Least Common Venue 33th Least Common Venue  \
Burgettstown        American Restaurant                    Food   
Charleroi                          Food      Italian Restaurant   
Donora                             Food             Pizza Place   
East Washington     American Restaurant             Pizza Place   
Eighty Four                 Wings Joint          Sandwich Place   

                Cluster Labels Shop Venue Food Venue Hip Places  
Burgettstown               4.0       10.0        9.0       19.0  
Charleroi                  4.0       23.0       10.0       33.0  
Donora                     4.0        7.0        5.0       12.0  
East Washington            4.0       24.0       16.0       40.0  
Eighty Four                4.0        4.0        1.0        5.0  

[5 rows x 48 columns]

Now we shall arrange the 13 cities in these common clusters in order of maximum number of hip places.

In [133]:
washington_hip1=washington_hip[['1st Least Common Venue','2nd Least Common Venue', '3rd Least Common Venue','Hip Places']].copy()
washington_hip2=washington_hip1.sort_values(['Hip Places'], ascending=[False])
washington_hip2

1st Least Common Venue 2nd Least Common Venue  \
Thompsonville       Fried Chicken Joint       Sushi Restaurant   
East Washington             Wings Joint      Korean Restaurant   
Charleroi           American Restaurant       Sushi Restaurant   
Burgettstown        Fried Chicken Joint       Sushi Restaurant   
Moninger            Fried Chicken Joint       Sushi Restaurant   
Finleyville         American Restaurant       Sushi Restaurant   
Donora              American Restaurant       Sushi Restaurant   
West Brownsville    American Restaurant       Sushi Restaurant   
Wickerham Manor     American Restaurant       Sushi Restaurant   
Speers              Fried Chicken Joint       Sushi Restaurant   
Eighty Four         American Restaurant       Sushi Restaurant   
Ellsworth           Fried Chicken Joint       Sushi Restaurant   
Elco                American Restaurant       Sushi Restaurant   

                 3rd Least Common Venue  Hip Places  
Thompsonville                Steakhouse        55.0  
East Washington      Italian Restaurant        40.0  
Charleroi                    Steakhouse        33.0  
Burgettstown                 Steakhouse        19.0  
Moninger                     Steakhouse        16.0  
Finleyville                  Steakhouse        15.0  
Donora                       Steakhouse        12.0  
West Brownsville             Steakhouse         8.0  
Wickerham Manor              Steakhouse         7.0  
Speers                       Steakhouse         6.0  
Eighty Four                  Steakhouse         5.0  
Ellsworth                    Steakhouse         5.0  
Elco                         Steakhouse         2.0

Thus we can see that Thompsonville is the busiest place with maximum shopping and eating places but lacks 'Fried Chicken Joint'. So our recommendation would be a Fried Chicken Joint in Thompsonville.

                             ------------------------XXXXXX------------------------